plasmapy/diagnostics/proton_radiography.py

```diff
@@ -300,6 +300,205 @@ def _log(self, msg):
+            mesh in the plane or tilt the mesh plane relative to the
+            source-detector axis. By default, `mesh_hdir` is set equal to
+            `detector_hdir` (see `detector_hdir` keyword in `__init__`).
+
+        mesh_vdir : `numpy.ndarray`, shape (3), optional
+            A unit vector (in Cartesian coordinates) defining the vertical
+            direction on the mesh plane. Modifying this vector can tilt the
+            mesh relative to the source-detector axis. By default, `mesh_vdir`
+            is defined to be perpendicular to `mesh_hdir` and the detector
+            plane normal (such that the mesh is parallel to the detector plane).
+
+        Raises
+        ------
+        ValueError
+            Raises a ValueError if the provided mesh location is not
+            between the source and the object grid.
+
+            Raises a ValueError if the specified grid removes ALL of the
+            particles.
+
+
+        Returns
+        -------
+
+        nremoved : int
+            The number of particles removed
+
+        """
+
+        location = _coerce_to_cartesian_si(location)
+        wire_radius = wire_diameter.si.value / 2
+
+        if not isinstance(extent, tuple):
+            extent = (extent,)
+
+        if len(extent) == 1:
+            radius = 0.5 * extent[0].si.value
+            width = extent[0].si.value
+            height = extent[0].si.value
+        elif len(extent) == 2:
+            radius = None
+            width = extent[0].si.value
+            height = extent[1].si.value
+        else:
+            raise ValueError(
+                "extent must be a tuple of 1 or 2 elements, but "
+                f"{len(extent)} elements were provided."
+            )
+
+        if not isinstance(nwires, tuple):
+            nwires = (nwires,)
+
+        if len(nwires) != 2:
+            nwires = (nwires[0], nwires[0])
+
+        # If no hdir/vdir is specified, calculate a default value
+        # If one is specified, make sure it is normalized
+        if mesh_hdir is None:
+            # Re-calculate the default here, in case the user
+            # specified a different det_hdir
+            mesh_hdir = self._default_detector_hdir()
+        else:
+            mesh_hdir = mesh_hdir / np.linalg.norm(mesh_hdir)
+
+        if mesh_vdir is None:
+            mesh_vdir = np.cross(mesh_hdir, self.det_n)
+            mesh_vdir = -mesh_vdir / np.linalg.norm(mesh_vdir)
+        else:
+            mesh_vdir = mesh_vdir / np.linalg.norm(mesh_vdir)
+
+        mesh_normal = np.cross(mesh_hdir, mesh_vdir)
+
+        # Raise exception if mesh is AFTER the field grid
+        if np.linalg.norm(location - self.source) > np.linalg.norm(self.source):
+            raise ValueError(
+                f"The specified mesh location, {location},"
+                "is not between the source and the origin."
+            )
+
+        # Calculate the times required to evolve each particle to the mesh
+        # plane
```

Comment:
So at this point you stop defining the geometry of this wire mesh and start to do stuff with particles. Doesn't this suggest we should break this up into two functions/methods/whatever? This one seems like it's doing too much.

Also, suppose I ran `sim.run` first, then added the wire mesh? What would happen then?

Reply:
Or suppose my experimental set up had two meshes one after the other?

I think we might want to have a list of (representations of physical) objects for particles to interact with. The mesh is one, the grid is another, and then the detector is a third one. Given their spatial ordering, we could then iterate over them.

Reply:
So we could replace part of this function with an `coast_to_plane(center, hdir, vdir)` function. However it wouldn't replace `coast_to_grid` (which doesn't assume the grid is rectangular) or `coast_to_detector` (which also handles removal of particles going the wrong way. We could potentially refactor the particle deletion part of `coast_to_detector` out, and then have `coast_to_plane` and `remove_particles_heading_away`?

I could see having a list of meshes? Although I've never seen an experiment that used two, it is in theory possible.

Reply:
I'll draft a refactor sketch later to show you what I mean here - currently a little swamped!